### ISSUE TITLE
go get with branch results in 500 error

### ISSUE THREAD

**Issue by rogpeppe** [gomods/athens#950]
**Describe the bug**
If I use `go get $module@$branchname`, I get a 500 response from the Athens proxy.

**Error Message**

```console
go get github.com/juju/errors@master: unexpected status (http://localhost:3000/github.com/juju/errors/@v/master.info): 500 Internal Server Error
```

**To Reproduce**

```console
% cd /tmp
% mkdir m
% cd m
% go mod init example.com/foo
go: creating new go.mod: module example.com/foo
% cat main.go
package main

import _ "github.com/juju/errors"

func main() {
}
% cat go.mod
module example.com/foo

go 1.12

require github.com/juju/errors v0.0.0-20180806074554-22422dad46e1
% echo $GOPROXY
http://localhost:3000
% go get github.com/juju/errors@master
go: finding github.com/juju/errors master
go: finding github.com/juju master
go: finding github.com master
go get github.com/juju/errors@master: unexpected status (http://localhost:3000/github.com/juju/errors/@v/master.info): 500 Internal Server Error
```

**Expected behavior**
I would expect this command to fetch the latest commit from the master branch of `github.com/juju/errors`.

**Environment (please complete the following information):**
 - OS: Linux 64 bit
 - Go version : go version devel +22dbc96d8a Tue Nov 27 09:26:16 2018 +0000 linux/amd64
 - Buffalo Version : no sure how to find this out
 - Proxy version : gomods/athens:v0.2.0 
 - Storage (fs/mongodb/s3 etc.) : fs

**Docker logs:**

I observed the following docker log messages when running the above script:

```console
handler: GET /github.com/juju/errors/@v/list/ [200]
time="2018-11-27T12:37:57Z" level=error msg="not in cache" http-method=GET http-path=/github.com/juju/errors/@v/master.info http-url=/github.com/juju/errors/@v/master.info kind="Internal Server Error" module=github.com/juju/errors operation=download.VersionInfoHandler ops="[download.VersionInfoHandler pool.Info protocol.Info stash.Pool stasher.Stash stasher.fetchModule goGetFetcher.Fetch module.downloadModule]" version=master
handler: GET /github.com/juju/errors/@v/master.info [500]
time="2018-11-27T12:37:57Z" level=error msg="invalid version \"master\"" http-method=GET http-path=/github.com/juju/@v/master.info http-url=/github.com/juju/@v/master.info kind="Internal Server Error" module=github.com/juju operation=download.VersionInfoHandler ops="[download.VersionInfoHandler pool.Info protocol.Info stash.Pool stasher.Stash stasher.fetchModule goGetFetcher.Fetch module.downloadModule]" version=master
handler: GET /github.com/juju/@v/master.info [500]
time="2018-11-27T12:37:58Z" level=error msg="invalid version \"master\"" http-method=GET http-path=/github.com/@v/master.info http-url=/github.com/@v/master.info kind="Internal Server Error" module=github.com operation=download.VersionInfoHandler ops="[download.VersionInfoHandler pool.Info protocol.Info stash.Pool stasher.Stash stasher.fetchModule goGetFetcher.Fetch module.downloadModule]" version=master
handler: GET /github.com/@v/master.info [500]
```

**Comment by michalpristas** [comment]
Related #949

**Comment by rogpeppe** [comment]
Note that `go get github.com/juju/errors` did not update to the latest version either.

**Comment by marpio** [comment]
@rogpeppe this is a [known limitation](https://docs.gomods.io/intro/known-limitations/) Documentation mentions commit hash but it applies also to branches.
What you can do is `go get` this module without using proxy (so GOPROXY="") 
This will add something like `require github.com/juju/errors v0.0.0-20181118221551-089d3ea4e4d5 // indirect` to the `go.mod` file. The build using proxy will work because now the proxy will use the pseudo-version ` v0.0.0-20181118221551-089d3ea4e4d5`.
Ref https://github.com/golang/go/issues/27947

**Comment by rogpeppe** [comment]
FWIW the known limitation page says:

> When you want to include a new library in your program by “branch”, just do go get github.com/my/pkg@master. This will put a real semver pseudo version inside your go.mod file. From that point on, you can use the GOPROXY to get the same exact version.

which sounds like the branch code works.

**Comment by marpio** [comment]
You're right, we should make the docs more explicit/clear.
`go get github.com/juju/errors@master` wont work if `$GOPROXY` env var is set. It will work though, if it's **not** set.
Once you have a record like this `require github.com/juju/errors v0.0.0-20180806074554-22422dad46e1` in your `go.mod`, go will use the `pseudo version` while building and so `athens` will get requests like `/github.com/juju/@v/v0.0.0-20180806074554-22422dad46e1.info` which are all fine.
So to summarize:
- `go get github.com/juju/errors@master` will work only if `GOPROXY` is **not** set.
- `go get github.com/juju/errors@v0.0.0-20180806074554-22422dad46e1` will work with `GOPROXY` set

**Comment by arschles** [comment]
Another useful addition would be to include the error message in the response body. When you run `go get github.com/juju/errors@master` against an Athens server, here's what the server logs look like:

```handler: GET /github.com/juju/errors/@v/master.info [500]
time="2018-12-06T19:14:13Z" level=error msg="invalid version \"master\"" http-method=GET http-path=/github.com/juju/@v/master.info http-url=/github.com/juju/@v/master.info kind="Internal Server Error" module=github.com/juju operation=download.VersionInfoHandler ops="[download.VersionInfoHandler pool.Info protocol.Info stash.Pool stasher.Stash stasher.fetchModule goGetFetcher.Fetch module.downloadModule]" version=master
handler: GET /github.com/juju/@v/master.info [500]
time="2018-12-06T19:14:13Z" level=error msg="invalid version \"master\"" http-method=GET http-path=/github.com/@v/master.info http-url=/github.com/@v/master.info kind="Internal Server Error" module=github.com operation=download.VersionInfoHandler ops="[download.VersionInfoHandler pool.Info protocol.Info stash.Pool stasher.Stash stasher.fetchModule goGetFetcher.Fetch module.downloadModule]" version=master
```

We could return the `msg` (`invalid version "master"`) to the client. That would be a good start for us

Maybe in the future we can match against this string on the server and return a 404 if we find it.....

**Comment by marpio** [comment]
We're now able to go get by branch!

**Comment by stubbornTanzhe** [comment]
> We're now able to go get by branch!

is this feature merged to master branch? I checked athen docker but found not supported yet.

**Comment by marpio** [comment]
@stubbornTanzhe 
Please try again, now that https://github.com/gomods/athens/pull/1063 is merged - this fix requires go 1.12 to work properly.
